config/hostcredentials/transform.go

```diff
@@ -0,0 +1,76 @@
+// Package hostcredentials implements a TransformationProvider that fetches
+// credentials from the (oddly named) `host-secrets` service and replaces
+// objects of the form: {$hostcredentials: true} with the credentials.
+//
+// The configuration value `hostCredentialsUrls` must be a list of URLS for
+// host-secrets servers, which will be tried in order repeatedly until one
+// succeeds.
+//
+// Note that this transform will need to run before the "hostsecrets" transform.
+package hostcredentials
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/taskcluster/taskcluster-worker/config"
+)
+
+type provider struct{}
+
+func init() {
+	config.Register("hostcredentials", provider{})
+}
+
+func (provider) Transform(cfg map[string]interface{}) error {
+	urls, ok := cfg["hostCredentialsUrls"].([]interface{})
+	if !ok {
+		return errors.New(
+			"Expected 'hostCredentialsUrls' property to hold a list of URLs for host-secrets services")
+	}
+
+	var creds interface{}
```

Comment:
Ah, I see what pmoore means with validation...

Yeah I would do:
```go
var creds struct {
  ClientID string `json:"clientId"`
  AccessToken string `json:"accessToken"`
  Certificate string `json:"certificate"`
}
```

even if it means that we later have to do:
```go
let retval := map[string]string{
  "clientId": creds.ClientID,
  "accessToken": creds.AccessToken,
}
if creds.Certificate != "" {
  retval["certificate"] = creds.Certificate
}
return retval
```

This way:
A) We ensure that we don't inject config values we don't want from a buggy server. Which I think is what @petemoore wanted to guard against (it's not JSON schema validation, but does provide some minimum of sanity).
B) We always return a clone of the cache...